src/main/java/slimeknights/tconstruct/smeltery/tileentity/AbstractCastingTileEntity.java

```diff
@@ -0,0 +1,232 @@
+package slimeknights.tconstruct.smeltery.tileentity;
+
+import com.google.common.collect.Lists;
+import lombok.Getter;
+import lombok.Setter;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.inventory.ISidedInventory;
+import net.minecraft.inventory.container.Container;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipeType;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.particles.ParticleTypes;
+import net.minecraft.tileentity.ITickableTileEntity;
+import net.minecraft.tileentity.TileEntityType;
+import net.minecraft.util.Direction;
+import net.minecraft.util.SoundCategory;
+import net.minecraft.util.SoundEvents;
+import net.minecraft.util.text.TranslationTextComponent;
+import net.minecraft.world.server.ServerWorld;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.items.ItemHandlerHelper;
+import net.minecraftforge.items.wrapper.SidedInvWrapper;
+import slimeknights.mantle.tileentity.InventoryTileEntity;
+import slimeknights.tconstruct.library.fluid.FluidTankAnimated;
+import slimeknights.tconstruct.library.network.TinkerNetwork;
+import slimeknights.tconstruct.library.smeltery.CastingFluidHandler;
+import slimeknights.tconstruct.smeltery.network.FluidUpdatePacket;
+import slimeknights.tconstruct.smeltery.recipe.AbstractCastingRecipe;
+import slimeknights.tconstruct.smeltery.recipe.ICastingInventory;
+import slimeknights.tconstruct.smeltery.recipe.TileCastingWrapper;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+
+public class AbstractCastingTileEntity extends InventoryTileEntity implements ITickableTileEntity, ISidedInventory, FluidUpdatePacket.IFluidPacketReceiver {
+  public static final int INPUT = 0;
+  protected static final int OUTPUT = 1;
+  public FluidTankAnimated tank;
+  public LazyOptional<CastingFluidHandler> holder = LazyOptional.of(() -> new CastingFluidHandler(this, tank));
+  private final ICastingInventory crafting;
+  protected int timer;
+  @Getter @Setter
+  private Fluid fluid;
```

Comment:
Might be better to put this field in the wrapper, like I did with the part builder with the material recipe. Reason being this is not technically the fluid of the TE, but the one used for finding recipes. May cause a mod using our "API" to call the wrong method.